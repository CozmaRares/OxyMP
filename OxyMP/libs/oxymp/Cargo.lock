# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "aho-corasick"
version = "1.1.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e60d3430d3a69478ad0993f19238d2df97c507009a52b3c10addcd7f6bcb916"
dependencies = [
 "memchr",
]

[[package]]
name = "bit-set"
version = "0.5.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0700ddab506f33b20a03b13996eccd309a48e5ff77d0d95926aa0210fb4e95f1"
dependencies = [
 "bit-vec",
]

[[package]]
name = "bit-vec"
version = "0.6.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "349f9b6a179ed607305526ca489b34ad0a41aed5f7980fa90eb03160b69598fb"

[[package]]
name = "change-case"
version = "0.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2cb9b85fd173fbe43d17902c970f23e5feda453526f4b28dbb4d2e688dffb92c"
dependencies = [
 "fancy-regex",
 "lazy_static",
 "regex",
]

[[package]]
name = "equivalent"
version = "1.0.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "877a4ace8713b0bcf2a4e7eec82529c029f1d0619886d18145fea96c3ffe5c0f"

[[package]]
name = "fancy-regex"
version = "0.3.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ae91abf6555234338687bb47913978d275539235fcb77ba9863b779090b42b14"
dependencies = [
 "bit-set",
 "regex",
]

[[package]]
name = "glob"
version = "0.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a8d1add55171497b4705a648c6b583acafb01d58050a51727785f0b2c8e0a2b2"

[[package]]
name = "hashbrown"
version = "0.15.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bf151400ff0baff5465007dd2f3e717f3fe502074ca563069ce3a6629d07b289"

[[package]]
name = "indexmap"
version = "2.7.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8c9c992b02b5b4c94ea26e32fe5bccb7aa7d9f390ab5c1221ff895bc7ea8b652"
dependencies = [
 "equivalent",
 "hashbrown",
]

[[package]]
name = "itoa"
version = "1.0.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d75a2a4b1b190afb6f5425f10f6a8f959d2ea0b9c2b1d79553551850539e4674"

[[package]]
name = "lazy_static"
version = "1.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bbd2bcb4c963f2ddae06a2efc7e9f3591312473c50c6685e1f298068316e66fe"

[[package]]
name = "memchr"
version = "2.7.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "78ca9ab1a0babb1e7d5695e3530886289c18cf2f87ec19a575a0abdce112e3a3"

[[package]]
name = "nom"
version = "8.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "df9761775871bdef83bee530e60050f7e54b1105350d6884eb0fb4f46c2f9405"
dependencies = [
 "memchr",
]

[[package]]
name = "oxymp"
version = "0.1.0"
dependencies = [
 "oxymp_lexer",
 "oxymp_lr",
 "oxymp_macro_utils",
 "oxymp_rd",
 "oxymp_tokens",
 "oxymp_util",
 "proc-macro2",
 "syn",
 "trybuild",
]

[[package]]
name = "oxymp_lexer"
version = "0.1.0"
dependencies = [
 "oxymp_macro_utils",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_lr"
version = "0.1.0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_macro_utils"
version = "0.1.0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_rd"
version = "0.1.0"
dependencies = [
 "change-case",
 "nom",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_tokens"
version = "0.1.0"
dependencies = [
 "change-case",
 "oxymp_macro_utils",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "oxymp_util"
version = "0.1.0"
dependencies = [
 "regex",
]

[[package]]
name = "proc-macro2"
version = "1.0.93"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "60946a68e5f9d28b0dc1c21bb8a97ee7d018a8b322fa57838ba31cc878e22d99"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.38"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0e4dccaaaf89514f546c693ddc140f729f958c247918a13380cccc6078391acc"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "regex"
version = "1.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b544ef1b4eac5dc2db33ea63606ae9ffcfac26c1416a2806ae0bf5f56b201191"
dependencies = [
 "aho-corasick",
 "memchr",
 "regex-automata",
 "regex-syntax",
]

[[package]]
name = "regex-automata"
version = "0.4.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "809e8dc61f6de73b46c85f4c96486310fe304c434cfa43669d7b40f711150908"
dependencies = [
 "aho-corasick",
 "memchr",
 "regex-syntax",
]

[[package]]
name = "regex-syntax"
version = "0.8.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2b15c43186be67a4fd63bee50d0303afffcef381492ebe2c5d87f324e1b8815c"

[[package]]
name = "ryu"
version = "1.0.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6ea1a2d0a644769cc99faa24c3ad26b379b786fe7c36fd3c546254801650e6dd"

[[package]]
name = "serde"
version = "1.0.217"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "02fc4265df13d6fa1d00ecff087228cc0a2b5f3c0e87e258d8b94a156e984c70"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.217"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5a9bf7cf98d04a2b28aead066b7496853d4779c9cc183c440dbac457641e19a0"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "serde_json"
version = "1.0.138"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d434192e7da787e94a6ea7e9670b26a036d0ca41e0b7efb2676dd32bae872949"
dependencies = [
 "itoa",
 "memchr",
 "ryu",
 "serde",
]

[[package]]
name = "serde_spanned"
version = "0.6.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "87607cb1398ed59d48732e575a4c28a7a8ebf2454b964fe3f224f2afc07909e1"
dependencies = [
 "serde",
]

[[package]]
name = "syn"
version = "2.0.98"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "36147f1a48ae0ec2b5b3bc5b537d267457555a10dc06f3dbc8cb11ba3006d3b1"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "target-triple"
version = "0.1.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "42a4d50cdb458045afc8131fd91b64904da29548bcb63c7236e0844936c13078"

[[package]]
name = "termcolor"
version = "1.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "06794f8f6c5c898b3275aebefa6b8a1cb24cd2c6c79397ab15774837a0bc5755"
dependencies = [
 "winapi-util",
]

[[package]]
name = "toml"
version = "0.8.20"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cd87a5cdd6ffab733b2f74bc4fd7ee5fff6634124999ac278c35fc78c6120148"
dependencies = [
 "serde",
 "serde_spanned",
 "toml_datetime",
 "toml_edit",
]

[[package]]
name = "toml_datetime"
version = "0.6.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0dd7358ecb8fc2f8d014bf86f6f638ce72ba252a2c3a2572f2a795f1d23efb41"
dependencies = [
 "serde",
]

[[package]]
name = "toml_edit"
version = "0.22.24"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "17b4795ff5edd201c7cd6dca065ae59972ce77d1b80fa0a84d94950ece7d1474"
dependencies = [
 "indexmap",
 "serde",
 "serde_spanned",
 "toml_datetime",
 "winnow",
]

[[package]]
name = "trybuild"
version = "1.0.103"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b812699e0c4f813b872b373a4471717d9eb550da14b311058a4d9cf4173cbca6"
dependencies = [
 "glob",
 "serde",
 "serde_derive",
 "serde_json",
 "target-triple",
 "termcolor",
 "toml",
]

[[package]]
name = "unicode-ident"
version = "1.0.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "00e2473a93778eb0bad35909dff6a10d28e63f792f16ed15e404fca9d5eeedbe"

[[package]]
name = "winapi-util"
version = "0.1.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cf221c93e13a30d793f7645a0e7762c55d169dbb0a49671918a2319d289b10bb"
dependencies = [
 "windows-sys",
]

[[package]]
name = "windows-sys"
version = "0.59.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1e38bc4d79ed67fd075bcc251a1c39b32a1776bbe92e5bef1f0bf1f8c531853b"
dependencies = [
 "windows-targets",
]

[[package]]
name = "windows-targets"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9b724f72796e036ab90c1021d4780d4d3d648aca59e491e6b98e725b84e99973"
dependencies = [
 "windows_aarch64_gnullvm",
 "windows_aarch64_msvc",
 "windows_i686_gnu",
 "windows_i686_gnullvm",
 "windows_i686_msvc",
 "windows_x86_64_gnu",
 "windows_x86_64_gnullvm",
 "windows_x86_64_msvc",
]

[[package]]
name = "windows_aarch64_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "32a4622180e7a0ec044bb555404c800bc9fd9ec262ec147edd5989ccd0c02cd3"

[[package]]
name = "windows_aarch64_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "09ec2a7bb152e2252b53fa7803150007879548bc709c039df7627cabbd05d469"

[[package]]
name = "windows_i686_gnu"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e9b5ad5ab802e97eb8e295ac6720e509ee4c243f69d781394014ebfe8bbfa0b"

[[package]]
name = "windows_i686_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0eee52d38c090b3caa76c563b86c3a4bd71ef1a819287c19d586d7334ae8ed66"

[[package]]
name = "windows_i686_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "240948bc05c5e7c6dabba28bf89d89ffce3e303022809e73deaefe4f6ec56c66"

[[package]]
name = "windows_x86_64_gnu"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "147a5c80aabfbf0c7d901cb5895d1de30ef2907eb21fbbab29ca94c5b08b1a78"

[[package]]
name = "windows_x86_64_gnullvm"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "24d5b23dc417412679681396f2b49f3de8c1473deb516bd34410872eff51ed0d"

[[package]]
name = "windows_x86_64_msvc"
version = "0.52.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "589f6da84c646204747d1270a2a5661ea66ed1cced2631d546fdfb155959f9ec"

[[package]]
name = "winnow"
version = "0.7.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "59690dea168f2198d1a3b0cac23b8063efcd11012f10ae4698f284808c8ef603"
dependencies = [
 "memchr",
]
