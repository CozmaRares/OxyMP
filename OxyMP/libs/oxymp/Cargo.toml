[package]
name = "oxymp"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[features]
default = ["rd"]
full = ["rd", "lr"]
rd = []
lr = []

[dependencies]
proc-macro2 = "1.0"
syn = { version = "2.0", features = ["extra-traits", "full"] }
trybuild = "1.0.103"
oxymp_util = { version = "0.1.0", path = "../oxymp-util" }
oxymp_lexer = { version = "0.1.0", path = "../oxymp-lexer" }
oxymp_tokens = { version = "0.1.0", path = "../oxymp-tokens" }
oxymp_rd = { version = "0.1.0", path = "../oxymp-rd" }
oxymp_lr = { version = "0.1.0", path = "../oxymp-lr" }
