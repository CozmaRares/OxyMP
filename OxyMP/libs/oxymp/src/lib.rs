use oxymp_lexer::derive_lexer_impl;
use oxymp_tokens::derive_tokens_impl;
use quote::quote;

#[cfg(feature = "rd")]
use oxymp_rd::derive_rd_impl;

#[cfg(feature = "lr")]
use oxymp_lr::derive_lr_impl;

#[proc_macro_derive(Tokens, attributes(exact, regex))]
pub fn derive_tokens(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_tokens_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[proc_macro_derive(Lexer, attributes(tokens, skip))]
pub fn derive_lexer(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_lexer_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[cfg(feature = "rd")]
#[proc_macro_derive(RDParser, attributes(grammar, tokens, grammar_tokens))]
pub fn derive_rd(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_rd_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

#[cfg(feature = "lr")]
#[proc_macro_derive(LRParser, attributes(grammar, tokens, grammar_tokens))]
pub fn derive_lr(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match derive_lr_impl(input.into()) {
        Ok(o) => o,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

// All the credit goes to
// experiment & starting point - https://github.com/faassen/outer_macro_pattern
// the *only* video explaining the outer macro pattern  - https://www.youtube.com/watch?v=aEWbZxNCH0A
use syn::{parse_macro_input, spanned::Spanned};

#[proc_macro_attribute]
pub fn outer(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let item_mod = parse_macro_input!(item as syn::ItemMod);
    let mut structs = Vec::new();
    let mut functions = Vec::new();
    eprintln!("{:#?}", &item_mod.content);

    if let Some((_, items)) = &item_mod.content {
        for item in items {
            match item {
                syn::Item::Struct(item_struct) => {
                    if has_inner_attribute(&item_struct.attrs) {
                        structs.push(item_struct);
                    }
                }
                syn::Item::Fn(item_fn) => {
                    if has_inner_attribute(&item_fn.attrs) {
                        functions.push(item_fn);
                    }
                }
                _ => {}
            }
        }
        // just print the structs and functions for now
        for item_struct in &structs {
            eprintln!("struct: {:#?}", item_struct.ident);
        }
        for item_fn in &functions {
            eprintln!("function: {:#?}", item_fn.sig.ident);
        }
    } else {
        // marking a module with #[outer] only makes sense if it has content
        // generate proper syn error with span
        return syn::Error::new(
            item_mod.span(),
            "module marked with #[outer] must have content",
        )
        .to_compile_error()
        .into();
    }

    quote! {
        #item_mod
    }
    .into()
}

fn has_inner_attribute(attributes: &[syn::Attribute]) -> bool {
    for attribute in attributes {
        if attribute.path().is_ident("inner") {
            return true;
        }
    }
    false
}

/// Mark a struct or function with #[inner] to be found by the #[outer] macro
#[proc_macro_attribute]
pub fn inner(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}
