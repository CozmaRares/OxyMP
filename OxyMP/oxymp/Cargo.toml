[package]
name = "oxymp"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[features]
default = ["rd"]
full = ["rd", "lr"]
rd = []
lr = []

[dependencies]
proc-macro2 = "1.0"
syn = { version = "2.0", features = ["extra-traits", "full"] }
quote = "1.0.38"
change-case = "0.2.0"

[dev-dependencies]
trybuild = "1.0.103"
