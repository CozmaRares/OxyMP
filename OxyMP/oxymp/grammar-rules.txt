/*
 * Derived from https://www.w3.org/TR/REC-xml/#sec-notation
 *
 * I took the liberty of adding a method for defining regex patterns.
 * The construct `r/.../` denotes a regex pattern.
**/

gramamr_rule ::= IDENT '=' choice

choice ::= list ('|' list)*

list ::= list_item+

list_item ::= (group | STRING | IDENT) ('?')?

group ::= '(' choice ')'

IDENT -> Rust identifier
STRING -> Rust string literal
