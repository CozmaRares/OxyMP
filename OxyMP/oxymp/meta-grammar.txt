/**
 * Derived from https://www.w3.org/TR/REC-xml/#sec-notation
**/

gramamr_rule ::= IDENT '=' choice

choice ::= list ('|' list)*

list ::= list_item+

list_item ::= (group | STRING | IDENT) ('?')?

group ::= '(' choice ')'

IDENT -> Rust identifier
STRING -> Rust string literal
