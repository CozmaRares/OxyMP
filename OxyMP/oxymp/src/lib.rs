// DOC: include this
// All the credit goes to
// experiment & starting point - https://github.com/faassen/outer_macro_pattern
// the *only* video explaining the outer macro pattern  - https://www.youtube.com/watch?v=aEWbZxNCH0A

mod data;
mod utils;

use data::{process_module, MacroData};
use quote::quote;
use syn::spanned::Spanned;

fn oxymp_impl(item: proc_macro::TokenStream) -> syn::Result<proc_macro2::TokenStream> {
    let mut item_mod: syn::ItemMod = match syn::parse(item) {
        Ok(o) => o,
        Err(e) => {
            return Err(syn::Error::new(
                e.span(),
                "Could not parse module. Make sure the attribute is applied to a module.",
            ))
        }
    };

    let Some((brace, items)) = item_mod.content else {
        return Err(syn::Error::new(
            item_mod.span(),
            "Module marked with #[oxymp] must have content.",
        ));
    };

    let (data, mut items) = process_module(items, &item_mod.ident)?;

    eprintln!("{:#?}", data);

    let generated_items = generate(data);
    items.extend(generated_items);

    item_mod.content = Some((brace, items));

    Ok(quote! {
        #item_mod
    })
}

#[proc_macro_attribute]
pub fn oxymp(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    match oxymp_impl(item) {
        Ok(tokens) => tokens,
        Err(e) => e.to_compile_error(),
    }
    .into()
}

fn generate(_data: MacroData) -> Vec<syn::Item> {
    vec![]
}
