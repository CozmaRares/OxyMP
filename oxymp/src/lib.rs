// All APIs need docs!
// #![deny(missing_docs)]
// TODO: rustdoc compatible documentation
// TODO: see if you can include structs only when generating docs
// to be used to document the generated data structures

// TODO: reorganize the order of functions and data structures in all the modules

#[macro_use]
mod macros;

#[cfg(debug_assertions)]
#[macro_use]
mod dbg_macros;

mod automata;
mod data;
mod generate;
mod grammar;
mod idents;
mod range;
mod symbols;
mod utils;

use data::process_module;
use generate::generate;
use syn::spanned::Spanned;

/// A procedural macro attribute used to generate a lexer and recursive descent
/// parser.
///
/// This macro simplifies the process of creating a parser for a language by:
///
/// * Generating a lexer based on regular expressions and exact matches defined
/// within a `Tokens` enum.
/// * Generating a recursive descent parser based on a grammar defined via the
/// `grammar` attribute.
///
/// The `oxymp` attribute must be applied to a Rust module (`mod`).
/// This module should contain the definitions for:
///
/// * Only one token enum marked with `#[oxymp::Tokens]`.
/// * Zero or more lexer modules marked with `#[oxymp::Lexer]`.
/// * Zero or more recursive descent parser modules marked with
/// `#[oxymp::RDParser]`.
///
/// **Example**
///
/// ```rust
/// #[oxymp::oxymp]
/// mod language {
///     use std::num::ParseFloatError;
///
///     pub fn match_number(input: &str) -> Result<f64, ParseFloatError> {
///         input.parse()
///     }
///
///     #[derive(Debug)]
///     #[oxymp::Tokens]
///     pub enum Tok {
///         #[regex(r"\d+(\.\d+)?", match_number)]
///         Number(f64),
///
///         #[exact(r"\+")]
///         Plus,
///
///         #[exact("-")]
///         Minus,
///     }
///
///     impl std::fmt::Display for Tok {
///         fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
///             match self {
///                 Tok::Number(n) => write!(f, "{}", n),
///                 Tok::Plus => write!(f, "+"),
///                 Tok::Minus => write!(f, "-"),
///             }
///         }
///     }
///
///     #[oxymp::Lexer]
///     #[skip(r"[ \t]+")]
///     #[error(ParseFloatError)]
///     pub mod lexer {}
///
///     #[oxymp::RDParser]
///     #[grammar(T = Number (r"\+" | "-") Number)]
///     pub mod rd_parser {}
/// }
/// use language::*;
///
/// fn main() -> Result<(), Box<dyn std::error::Error>> {
///     let input = "1 + 2";
///     println!("{:#?}", input);
///
///     let tokens = lexer::tokenize(input)?;
///     println!("{:#?}", tokens);
///
///     let (_inp, ast) = rd_parser::T(tokens.into())?;
///     println!("{:#?}", ast);
///
///     Ok(())
/// }
/// ```
///
/// **Token Definition (`#[oxymp::Tokens]`)**
///
/// The `Tokens` enum defines the set of tokens that the lexer will recognize.
/// Each variant of the enum represents a different token type. The following
/// attributes can be applied to token variants:
///
/// * `#[regex(pattern, match_function)]`:  Matches the token based on a
///   regular expression. The `pattern` is a regular expression string.
///   The `match_function` is a function that takes a `&str` (the matched
///   text) as input and returns a `Result<T, E>`, where `T` is the type of
///   the token's value and `E` is type that implements the `std::error::Error`
///   trait.
/// * `#[exact(pattern)]`: Matches the token exactly against the given
///   `pattern` string.
///
/// A module marked with `#[oxymp]` must contain only one token enum.
///
/// An enum marked with `#[oxymp::Tokens]` must implement the `std::fmt::Display`
/// trait.
///
/// **Lexer Module (`#[oxymp::Lexer]`)**
///
/// The `Lexer` module configures the lexer. The following attributes can be
/// applied to the `Lexer` module:
///
/// * `#[skip(pattern)]`: Specifies a regular expression that matches text
///   to be skipped (e.g., whitespace).
/// * `#[error(error_type)]`: Specifies the error type to use when tokenizing.
///
/// The lexer will generate:
///
/// * An `Error` eunm that is used to report errors during tokenization. It has
///   two variants:
///   * `Native` => the error was caused by the lexer itself when an unexpedted
///   character was encountered.
///   * `UserMatcher` => the error was caused by functions defined by the user
///   for a `#[regex(...)]` token definition.
/// * A `tokenize` function that takes a `&str` as input and returns a
///   `Result<Vec<Token>, Error>`, where `Token` is the token enum defined with
///   `#[oxymp::Tokens]` and `Error` is the lexer's Error type.
///
/// A module marked with `#[oxymp]` *may* contain *zero* or *more* lexer modules.
///
/// **Parser Module (`#[oxymp::RDParser]`)**
///
/// The `RDParser` module configures the recursive descent parser.
///
/// To define a production rule, the following attribute can be applied to the
/// `RDParser` module: `#[grammar(ProductionName = ...)]`
///
/// * The `ProductionName` is the name of the production (e.g., `T` in the example).
/// * The right-hand side of the production specifies the sequence of tokens and/or
/// other productions that make up the production.
///   * *Token names* used in the grammar *must* match the variant names in the
///   `Tokens` enum or, in case of `#[exact(pattern)]` definition, the *pattern*
///   can also be used.
///   * A sequence can be indicated by `A B C`, where `A`, `B`, and `C` are tokens
///   or productions.
///   * Choice can be indicated by `sequence A | sequence B`.
///   * Optionality can be indicated by `sequence?`.
///
/// Token names used in the grammar
/// *must* match the variant names in the `Tokens` enum. `Choice` can be indicated
/// by `sequence A | sequence B`.
///
/// The parser module will generate:
///
/// * An `Input` struct that is used to pass the input to the parser. The
/// `From<Vec<Token>>` trait is implemented for this struct.
/// * An `ErrorKind` eunm that has the following variants:
///   * `UnexpectedEOf` => the parser encountered the end of input.
///   * `UnexpectedToken` => the parser encountered an unexpected token.
///   * `ChoiceFailed` => the parser failed to match a choice production.
///     * `rule_name` => the name of the rule that failed.
///     * `choice_idx` => the index of the choice that failed (starting from 1).
///     * `causes` => a vector of errors that were encountered while trying to
///       parse each branch of the choice.
/// * An `Error` struct that is used to report errors during parsing. It has:
///   * `kind` => the error kind.
///   * `input` => the input that caused the error.
///   * `cursor` => the position of the token that caused the error.
///   * `expected` => a vector of expected token names.
/// * A function for each production rule defined with the `grammar` attribute.
///   These functions take a slice of tokens as input and return a
///   `Result<(/* remaining tokens */, AST node), Error>`, where `AST node` is a
///   struct generated based on the right-hand side of the grammar rule, and `Error`
///   is the parser's Error type.
/// * Generated AST structures:
///   * Rule productions are wrapped in a `Box` to avoid recursive types.
///   * Tokens are converted to the corresponding enum variants prepended with
///   the name of the tokens enum, e.g., `Tok::Number(n)` becomes `TokNumber(n)`.
///   * Sequences are converted to a tuple of the corresponding AST nodes.
///   * Option productions are wrapped in an `Option`.
///   * Choice productions generate an `enum` with each branch as a variant.
///     * The `enum` is prepended with the name of the rule and appended with its
///     index in the definition , e.g., choice `1` of rule `T` becomes `TChoice1`.
///     * Each branch of the choice is represented by its index in the definition,
///     e.g., branch `1` becomes `_1`.
///     * Combining the two, the first branch of the first choice of rule `T`
///     becomes `TChoice1::_1(...)`.
///
/// A module marked with `#[oxymp]` *may* contain *zero* or *more* recursive
/// descent parser modules.
///
/// A great tool to see the generated code is [dtolnay's cargo expand](https://github.com/dtolnay/cargo-expand)
#[proc_macro_attribute]
pub fn oxymp(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    if !attr.is_empty() {
        let msg = "Invalid 'oxymp' attribute. The attribute must not have any arguments, such as `#[oxymp]`,`#[oxymp::oxymp]` or `#[::oxymp::oxymp]`.";
        let toks = proc_macro2::TokenStream::from(attr);
        let span = toks.span();
        return syn::Error::new(span, msg).to_compile_error().into();
    }

    match syn::parse::<syn::ItemMod>(item)
        .and_then(process_module)
        .and_then(generate)
    {
        Ok(tokens) => tokens,
        Err(e) => e.to_compile_error(),
    }
    .into()
}
