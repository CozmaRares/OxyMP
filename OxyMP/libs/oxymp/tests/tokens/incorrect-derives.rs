use oxymp::Tokens;

// must have at least one variant
#[derive(Tokens)]
enum Tok {}

// can't be a struct
#[derive(Tokens)]
struct Tok2;

// can't be a union
#[derive(Tokens)]
union Tok3 {
    f1: u32,
    f2: f32,
}

fn main() {}
