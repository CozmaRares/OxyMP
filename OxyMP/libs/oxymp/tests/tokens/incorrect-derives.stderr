error: The token enum must contain at least one variant. Consider adding a variant with `#[exact("your exact string")]` or `#[regex("your regex", path::to::function)]`.
 --> tests/tokens/incorrect-derives.rs:5:6
  |
5 | enum Tok {}
  |      ^^^

error: Tokens can only be derived for enums. Make sure that `#[derive(Tokens)]` is applied to an enum.
 --> tests/tokens/incorrect-derives.rs:9:1
  |
9 | struct Tok2;
  | ^^^^^^

error: Tokens can only be derived for enums. Make sure that `#[derive(Tokens)]` is applied to an enum.
  --> tests/tokens/incorrect-derives.rs:13:1
   |
13 | union Tok3 {
   | ^^^^^
